File: src/scoring.rs
use crate::types::Market;
use chrono::Utc;

pub fn calculate_score(market: &Market, net_profit: f64) -> f64 {
    let factors = ScoreFactors {
        profit_score: normalize_profit(net_profit),
        liquidity_score: normalize_liquidity(market.liquidity),
        volume_score: normalize_volume(market.volume),
        execution_risk: calculate_execution_risk(market),
        time_decay: calculate_time_decay(market.end_time),
    };

    (factors.profit_score * 0.4)
        + (factors.liquidity_score * 0.25)
        + (factors.volume_score * 0.15)
        + (factors.execution_risk * 0.15)
        + (factors.time_decay * 0.05)
}

struct ScoreFactors {
    profit_score: f64,
    liquidity_score: f64,
    volume_score: f64,
    execution_risk: f64,
    time_decay: f64,
}

fn normalize_profit(profit: f64) -> f64 {
    1.0 / (1.0 + std::f64::consts::E.powf(-50.0 * (profit - 0.025)))
}

fn normalize_liquidity(liquidity: f64) -> f64 {
    (liquidity / 100_000.0).min(1.0).ln_1p()
}

fn normalize_volume(volume: f64) -> f64 {
    (volume / 1_000_000.0).min(1.0).ln_1p()
}

fn calculate_execution_risk(_market: &Market) -> f64 {
    1.0
}

fn calculate_time_decay(end_time: Option<chrono::DateTime<Utc>>) -> f64 {
    match end_time {
        None => 0.5,
        Some(end) => {
            let remaining = (end - Utc::now()).num_hours() as f64;
            if remaining < 0.0 {
                0.0
            } else {
                (remaining / 168.0).min(1.0)
            }
        }
    }
}

File: internal/scoring/scoring.go
package scoring

import (
	"math"
	"time"

	"github.com/ochanomizu/predmarket-scanner/pkg/types"
)

type ScoreFactors struct {
	ProfitScore    float64 `json:"profit_score"`
	LiquidityScore float64 `json:"liquidity_score"`
	VolumeScore    float64 `json:"volume_score"`
	ExecutionRisk  float64 `json:"execution_risk"`
	TimeDecay      float64 `json:"time_decay"`
}

func CalculateScore(market types.Market, netProfit float64) (float64, ScoreFactors) {
	factors := ScoreFactors{
		ProfitScore:    normalizeProfit(netProfit),
		LiquidityScore: normalizeLiquidity(market.Liquidity),
		VolumeScore:    normalizeVolume(market.Volume),
		ExecutionRisk:  1.0,
		TimeDecay:      calculateTimeDecay(market.EndTime),
	}

	score := (factors.ProfitScore * 0.4) +
		(factors.LiquidityScore * 0.25) +
		(factors.VolumeScore * 0.15) +
		(factors.ExecutionRisk * 0.15) +
		(factors.TimeDecay * 0.05)

	return score, factors
}

func normalizeProfit(profit float64) float64 {
	return 1.0 / (1.0 + math.Exp(-50.0*(profit-0.025)))
}

func normalizeLiquidity(liquidity float64) float64 {
	normalized := (liquidity / 100000.0)
	if normalized > 1.0 {
		normalized = 1.0
	}
	return math.Log1p(normalized)
}

func normalizeVolume(volume float64) float64 {
	normalized := (volume / 1000000.0)
	if normalized > 1.0 {
		normalized = 1.0
	}
	return math.Log1p(normalized)
}

func calculateTimeDecay(endTime *time.Time) float64 {
	if endTime == nil {
		return 0.5
	}

	remaining := endTime.Sub(time.Now()).Hours()
	if remaining < 0 {
		return 0.0
	}

	result := remaining / 168.0
	if result > 1.0 {
		result = 1.0
	}

	return result
}
